Web Automation: Switch to Frame should actually focus the frame

burg opened the pull request:
#### 189bef0d1cc76b0a8505b9f765f1666fc7cf2f1a
<pre>
Web Automation: Switch to Frame should actually focus the frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=293445">https://bugs.webkit.org/show_bug.cgi?id=293445</a>
&lt;<a href="https://rdar.apple.com/151870530">rdar://151870530</a>&gt;

Reviewed by Devin Rousso.

Focusing the frame upon switching is required by the WebDriver specification.

This is a partial fix for several existing test_no_browsing_context failures.
With this fix, the driver-side implementation may infer that a frame is no longer
valid if Automation.switchToBrowsingContext with a frame argument fails. This is used
to implement spec steps such as

&quot;if the current browsing context is no longer open, return no such window&quot;.

Prior to this change, switching to a detached frame as in the wpt fixture `closed_frame`
would *not* fail as expected, causing the driver to think that the frame reference is still
valid, even when it is not.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::switchToBrowsingContext):
If the command is targeted at a frame, perform the IPC to focus the frame in the frame&apos;s web process.

* Source/WebKit/UIProcess/WebPageProxy.cpp: Add new template instantation.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::focusFrame): Added. Make sure to check that
the page, frame, and the frame&apos;s page reference are all valid before trying to focus the frame.
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in: Add new IPC message.

Canonical link: <a href="https://commits.webkit.org/296164@main">https://commits.webkit.org/296164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fc5e9497fe57dc879cf1041889f1345610265df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17635 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112759 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58084 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27916 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35726 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81646 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110471 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22122 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62027 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21560 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57522 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91492 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/15104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115860 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34610 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25522 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90682 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34986 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93182 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90423 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23061 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35341 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13121 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30373 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34531 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40087 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34277 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37636 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

yury-s commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -465,14 +465,24 @@ void WebAutomationSession::switchToBrowsingContext(const Inspector::Protocol::Au
 
-    // FIXME: Should we also switch to the frame that frameHandle points to?
-    m_client->requestSwitchToPage(*this, *page, [page = Ref { *page }, callback = WTFMove(callback)]() {
+    m_client->requestSwitchToPage(*this, *page, [protectedThis = Ref { *this }, frameID, page = Ref { *page }, callback = WTFMove(callback)]() mutable {
```
I don't see any references to `this` in the callbacks, do you really need `protectedThis` here?

dcrousso replied:
yeah and also is the `mutable` actually needed?

ditto below

yury-s reviewed (commented):
Looks good to me, but I'm not very familiar with the classic web driver implementation.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -465,14 +465,24 @@ void WebAutomationSession::switchToBrowsingContext(const Inspector::Protocol::Au
+
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!WebFrameProxy::webFrame(frameID.value()), FrameNotFound);
+
+        page->sendWithAsyncReplyToProcessContainingFrameWithoutDestinationIdentifier(frameID, Messages::WebAutomationSessionProxy::FocusFrame(page->webPageIDInMainFrameProcess(), frameID.value()), WTF::CompletionHandler<void(std::optional<String>&&)> { [callback = WTFMove(callback)] (std::optional<String>&& optionalError) mutable {
```
NIT: can you just say `CompletionHandler { ... }` instead?  would be neat to be able to similar to how you can `Ref { ... }` without the type

burg replied:
I just tried this but it complains that it can't deduce the template type arguments. :|

burg replied:
Likewise the mutable seems to be needed because... I'm not quite sure really.

```
/app/webkit/WebKitBuild/GTK/Release/WTF/Headers/wtf/StdLibExtras.h: In instantiation of ‘constexpr typename std::remove_reference<_Arg>::type&& std::move(T&&) [with WTF::CheckMoveParameterTag <anonymous> = WTF::CheckMoveParameter; T = const WTF::Function<void(experimental::fundamentals_v3::expected<void, WTF::String>)>&; typename remove_reference<_Arg>::type = const WTF::Function<void(experimental::fundamentals_v3::expected<void, WTF::String>)>]’:
/app/webkit/Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:487:266:   required from here
/app/webkit/WebKitBuild/GTK/Release/WTF/Headers/wtf/StdLibExtras.h:839:51: error: static assertion failed: T is const qualified.
```

burg replied:
https://ews-build.webkit.org/#/builders/2/builds/115744

burg replied:
Okay, my current understanding is that because we WTFMove the `callback` out of the closure, the closures need to be marked `mutable` so that they are not inferred as `const`.

dcrousso commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -635,6 +636,35 @@ void WebAutomationSessionProxy::resolveParentFrame(WebCore::PageIdentifier pageI
+        return;
+    }
+
+    String frameNotFoundErrorType = Inspector::Protocol::AutomationHelpers::getEnumConstantValue(Inspector::Protocol::Automation::ErrorMessage::FrameNotFound);
```
NIT: i realize this is to avoid repeated code, but if it's more common for the error cases to not be reached, then having it out here is wasteful, so id move it into each of the `if` below (you also could just merge the `if` together to avoid writing it twice)

burg committed:
Web Automation: Switch to Frame should actually focus the frame
https://bugs.webkit.org/show_bug.cgi?id=293445
<rdar://151870530>

Reviewed by Devin Rousso.

Focusing the frame upon switching is required by the WebDriver specification.

This is a partial fix for several existing test_no_browsing_context failures.
With this fix, the driver-side implementation may infer that a frame is no longer
valid if Automation.switchToBrowsingContext with a frame argument fails. This is used
to implement spec steps such as

"if the current browsing context is no longer open, return no such window".

Prior to this change, switching to a detached frame as in the wpt fixture `closed_frame`
would *not* fail as expected, causing the driver to think that the frame reference is still
valid, even when it is not.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::switchToBrowsingContext):
If the command is targeted at a frame, perform the IPC to focus the frame in the frame's web process.

* Source/WebKit/UIProcess/WebPageProxy.cpp: Add new template instantation.

* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::focusFrame): Added. Make sure to check that
the page, frame, and the frame's page reference are all valid before trying to focus the frame.
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in: Add new IPC message.

Canonical link: https://commits.webkit.org/296164@main
